pkg/segment/query/querystatus.go

```diff
@@ -206,35 +287,23 @@ func StartQuery(qid uint64, async bool, cleanupCallback func()) (*RunningQuerySt
 		}()
 	}
 
-	runningState := &RunningQueryState{
-		startTime:         time.Now(),
-		StateChan:         stateChan,
-		cleanupCallback:   cleanupCallback,
-		rqsLock:           &sync.Mutex{},
-		isAsync:           async,
-		timeoutCancelFunc: timeoutCancelFunc,
-	}
-	allRunningQueries[qid] = runningState
-	return runningState, nil
+	return timeoutCancelFunc
 }
 
-// Removes reference to qid. If qid does not exist this is a noop
-func DeleteQuery(qid uint64) {
-	LogGlobalSearchErrors(qid)
+func runQuery(wsData WaitStateData) {
 	arqMapLock.Lock()
-	rQuery, ok := allRunningQueries[qid]
-	if ok {
-		delete(allRunningQueries, qid)
+	defer arqMapLock.Unlock()
+	if wsData.rQuery.isCancelled {
+		return
 	}
-	arqMapLock.Unlock()
 
-	if ok && !rQuery.isCancelled {
-		rQuery.timeoutCancelFunc()
+	allRunningQueries[wsData.qid] = wsData.rQuery
 
-		if rQuery.cleanupCallback != nil {
-			rQuery.cleanupCallback()
-		}
-	}
+	wsData.rQuery.timeoutCancelFunc = setupTimeoutCancelFunc(wsData.qid)
+	wsData.rQuery.startTime = time.Now()
+
+	wsData.rQuery.StateChan <- &QueryStateChanData{StateName: READY}
+	wsData.rQuery.StateChan <- &QueryStateChanData{StateName: RUNNING}
```

Comment:
What's the purpose of this READY state if we immediately send RUNNING afterwards? Is READY sent somewhere else as well?

Reply:
No, READY state is only meant to signal that query can now proceed. RUNNING state update is actually send over the websocket.

Reply:
Ah it looks like READY is supposed to be received by one function to actually start the query and RUNNING is received by the websocket handler (to send to the UI like you said); is that right? In that case, could there be a race condition where the websocket handler receives READY instead of the segexecution function receiving it, and then the query is never started?

Reply:
No it would never happen since websocket is only initialized after we call `ExecuteAsyncQueryForNewPipeline` we are blocked until we receive a READY state in this method.